### PR TITLE
Feature/allow scratch for fastq

### PR DESCRIPTION
**Pull request by jsnedecor**

### CODE REVIEW

File: zippy/modular_runner.py [hunk 1]
```diff
@@ -266,7 +270,15 @@ def get_output(self, sample):
         return {'fastq': samples_to_return}
 
     def get_dependencies(self, sample):
-        return self.bcl2fastq_task
+        return self.tasks
+
+    def _remove_undefined(self, workflowRunner, dependencies, input_dir):
```
Comment (quejebo) [inline zippy/modular_runner.py]: for consistency, _remove_undetermined

File: zippy/modular_runner.py [hunk 2]
```diff
@@ -278,9 +290,37 @@ def workflow(self, workflowRunner):
...
+            self.tasks.append(copy_label)
+            
+            remove_scratch_cmd = "rm -rf {scratch}".format(
+                    scratch=temp_folder)
+            workflowRunner.addTask("remove_scratch_{}".format(self.identifier), remove_scratch_cmd, dependencies=copy_label) #doesn't stop any downstream analysis.
+        else: 
+            bcl2fastq_wf = Bcl2Fastq(self.params.bcl2fastq_path, self.params.sample_path, self.params.self.output_dir, self.params.sample_sheet, args=args, max_job_cores=self.get_core_count(16))
+            bcl_label = workflowRunner.addWorkflowTask(self.identifier, bcl2fastq_wf, dependencies=dependencies)
+
+            remove_label = self._remove_undefined(workflowRunner, bcl_label, self.params.self.output_dir)
```
Comment (quejebo) [inline zippy/modular_runner.py]: for consistency, remove_task (same with the other 'labels')

**Comment by quejebo** [comment]
For cleaning up undetermined this looks good, and definitely belongs here.

For running on scratch, it looks like you run it in the scratch directory, and then copy everything -- is there any processing going on that I'm missing?  If not, I have a very simple copy folder runner that I've written.  A more generic way to do this would be to run bcl2fastq giving it a scratch path as the output folder, and then copying the folder to your final destination in a separate stage.  

Does that cover your use case?  If so, I'll polish up the copy runner and upload it.

**Comment by jsnedecor** [comment]
Yup, that would cover what I need, thank you!

**Comment by aawise** [comment]
A new build is pushed, adding the copy folder runner.  Here's an example of how to use it:

        {
            "input_dir": "/path/to/scratch/folder",
            "identifier": "copyfolder",
            "output_dir": "/path/to/long/term/storage",
            "stage": "copyfolder",
            "skip": false
        },

**Comment by jsnedecor** [comment]
Want me to edit out the scratch parts or do you want to give it a go?

**Comment by jsnedecor** [comment]
Edited out scratch references, in way too many commits, but hey.

**Comment by jsnedecor** [comment]
Hm. I just realized, using the copy runner doesn't really work since I don't have the scratch output directory till after I've started execution. 

[edit]
Actually, I figured out a workaround, nevermind.

[edit2] Nope, I haven't figured out a better method actually. I need to be able to know where the fastq files end up once I've copied them, and the copy runner doesn't retain that information. This doesn't really work.

**Comment by quejebo** [comment]
> Hm. I just realized, using the copy runner doesn't really work since I don't have the scratch output directory till after I've started execution.

Hmm... could you elaborate on this?  Is this a directory which is being created in the bcl2fastqrunner?

**Comment by jsnedecor** [comment]
So here's the heart of the problem: 

bcl2fastq is writing to an output directory on scratch. It does its thing and completes and everything is cool. 
copyfolder copies the files. 

bcl2fastq is defining its output as "{scratch}/{sample}_R1.fastq.gz" and "{scratch}/{sample}_R2.fastq.gz" so when I go to collect samples downstream, that's where it's looking. If I copy the files, I need to change the place that the downstream steps are looking for the fastq files. I tried to do some sort of nastiness where I copied and then recollected the fastq samples downstream, but then I've got three steps to run bcl2fastq instead of one and it kinda defeats the whole purpose of doing it through zippy because basically the minute I copy from one place to another I lose all the information about the output files. 

Basically it doesn't seem to make sense to do it in a separate step unless the copy step can take the output of collect_samples and then remap the paths, but that seems tricky and not really necessary considering I only need to write to scratch for this particular step.

**Comment by quejebo** [comment]
So that's an oversight on my part, since copyfolder should forward its outputs from the previous stage.

Something like this in the zippy config is the goal
{
"stage": "bcl2fastq",
"identifier": "bcl2fastq"
....
}
{
        "input_dir": "/path/to/scratch/folder",
        "identifier": "copyfolder",
        "output_dir": "/path/to/long/term/storage",
        "stage": "copyfolder",
        "previous_stage": "bcl2fastq",
        "skip": false
    },
    {
        "stage": "whatever_is_next",
        "previous_stage": "copyfolder"
        .....
    }

I'll fix that.

**Comment by jsnedecor** [comment]
Cool, that definitely works.

**Comment by quejebo** [comment]
I've created a branch copyfolder_passthrough that should work as described.  Can you take a look and make sure it's working for your use case?

**Comment by jsnedecor** [comment]
Would it be possible to just take the output from the previous stage and copy to output_dir (i.e., no input_dir)? We dynamically set output directories and it would be nice to not have to set the temporary scratch path  twice for two different steps. It's not a huge deal, but it would make life easier.
